File: src/csharp/pulse/cdm/patient/actions/SEChestCompressionForceScale.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

namespace Pulse.CDM
{
  public class SEChestCompressionForceScale : SEPatientAction
  {
    protected SEScalar0To1 force_scale;
    protected SEScalarTime force_period;

    public SEChestCompressionForceScale()
    {
      force_scale = null;
      force_period = null;
    }

    public override void Clear()
    {
      base.Clear();
      if (force_scale != null)
        force_scale.Invalidate();
      if (force_period != null)
        force_period.Invalidate();
    }

    public override bool IsValid()
    {
      return HasForceScale() && HasForcePeriod();
    }

    public bool HasForceScale()
    {
      return force_scale == null ? false : force_scale.IsValid();
    }
    public SEScalar0To1 GetForceScale()
    {
      if (force_scale == null)
        force_scale = new SEScalar0To1();
      return force_scale;
    }

    public bool HasForcePeriod()
    {
      return force_period == null ? false : force_period.IsValid();
    }
    public SEScalarTime GetForcePeriod()
    {
      if (force_period == null)
        force_period = new SEScalarTime();
      return force_period;
    }
  }
}

File: src/cpp/cdm/patient/actions/SEChestCompressionForceScale.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "stdafx.h"
#include "patient/actions/SEChestCompressionForceScale.h"
#include "properties/SEScalar0To1.h"
#include "properties/SEScalarTime.h"
#include "io/protobuf/PBPatientActions.h"

SEChestCompressionForceScale::SEChestCompressionForceScale(Logger* logger) : SEPatientAction(logger)
{
  m_ForceScale = nullptr;
  m_ForcePeriod = nullptr;
}

SEChestCompressionForceScale::~SEChestCompressionForceScale()
{
  SAFE_DELETE(m_ForceScale);
  SAFE_DELETE(m_ForcePeriod);
}

void SEChestCompressionForceScale::Clear()
{
  SEPatientAction::Clear();
  INVALIDATE_PROPERTY(m_ForceScale);
  INVALIDATE_PROPERTY(m_ForcePeriod);
}

void SEChestCompressionForceScale::Copy(const SEChestCompressionForceScale& src, bool preserveState)
{
  //if(preserveState) // Cache any state before copy,
  PBPatientAction::Copy(src, *this);
  //if(preserveState) // Put back any state
}

bool SEChestCompressionForceScale::IsValid() const
{
  return SEPatientAction::IsValid() && HasForceScale();
}

bool SEChestCompressionForceScale::IsActive() const
{
  if (!SEPatientAction::IsActive())
    return false;
  return !m_ForceScale->IsZero();
}
void SEChestCompressionForceScale::Deactivate()
{
  SEPatientAction::Deactivate();
  Clear();//No stateful properties
}

const SEScalar* SEChestCompressionForceScale::GetScalar(const std::string& name)
{
  if (name.compare("ForceScale") == 0)
    return &GetForceScale();
  if (name.compare("ForcePeriod") == 0)
    return &GetForcePeriod();
  return nullptr;
}

bool SEChestCompressionForceScale::HasForceScale() const
{
  return m_ForceScale == nullptr ? false : m_ForceScale->IsValid();
}
SEScalar0To1& SEChestCompressionForceScale::GetForceScale()
{
  if (m_ForceScale == nullptr)
    m_ForceScale = new SEScalar0To1();
  return *m_ForceScale;
}
double SEChestCompressionForceScale::GetForceScale() const
{
  if (m_ForceScale == nullptr)
    return SEScalar::dNaN();
  return m_ForceScale->GetValue();
}

bool SEChestCompressionForceScale::HasForcePeriod() const
{
  return m_ForcePeriod == nullptr ? false : m_ForcePeriod->IsValid();
}
SEScalarTime& SEChestCompressionForceScale::GetForcePeriod()
{
  if (m_ForcePeriod == nullptr)
    m_ForcePeriod = new SEScalarTime();
  return *m_ForcePeriod;
}
double SEChestCompressionForceScale::GetForcePeriod(const TimeUnit& unit) const
{
  if (m_ForcePeriod == nullptr)
    return SEScalar::dNaN();
  return m_ForcePeriod->GetValue(unit);
}

void SEChestCompressionForceScale::ToString(std::ostream &str) const
{
  str << "Patient Action : Chest Compression";
  if (HasComment())
    str << "\n\tComment: " << m_Comment;
  str << "\n\tForceScale: "; HasForceScale() ? str << *m_ForceScale : str << "NaN";
  str << "\n\tForcePeriod: "; HasForcePeriod() ? str << *m_ForcePeriod : str << "NaN";
  str << std::flush;
}